service/matching/handler/engine.go

```diff
@@ -1202,6 +1192,133 @@ func (e *matchingEngineImpl) emitInfoOrDebugLog(
 	}
 }
 
+func (e *matchingEngineImpl) shutDownNonOwnedTasklists() error {
+	if !e.config.EnableTasklistOwnershipGuard() {
+		return nil
+	}
+	noLongerOwned, err := e.getNonOwnedTasklists()
+	if err != nil {
+		return err
+	}
+
+	e.taskListsLock.Lock()
+	defer e.taskListsLock.Unlock()
+
+	for _, tl := range noLongerOwned {
+		// for each of the tasklists that are no longer owned, kick off the
+		// process of stopping them. The stopping process is IO heavy and
+		// can take a while, so do them in parallel so as to not hold the
+		// lock too long.
+		go func(tl tasklist.Manager) {
+
+			defer func() {
+				if r := recover(); r != nil {
+					e.logger.Error("panic occurred while trying to shut down tasklist", tag.Dynamic("recovered-panic", r))
+				}
+			}()
+
+			e.logger.Info("shutting down tasklist preemptively because they are no longer owned by this host",
+				tag.WorkflowTaskListName(tl.TaskListID().GetName()),
+				tag.WorkflowDomainID(tl.TaskListID().GetDomainID()),
+				tag.Dynamic("tasklist-debug-info", tl.String()),
+			)
+			tl.Stop()
+		}(tl)
+	}
+	return nil
+}
+
+func (e *matchingEngineImpl) getNonOwnedTasklists() ([]tasklist.Manager, error) {
+	if !e.config.EnableTasklistOwnershipGuard() {
+		return nil, nil
+	}
+
+	toShutDown := make([]tasklist.Manager, 0)
+
+	e.taskListsLock.RLock()
+	defer e.taskListsLock.RUnlock()
+
+	self, err := e.membershipResolver.WhoAmI()
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup self im membership: %w", err)
+	}
+
+	for tl, manager := range e.taskLists {
+		taskListOwner, err := e.membershipResolver.Lookup(service.Matching, tl.GetName())
+		if err != nil {
+			return nil, fmt.Errorf("failed to lookup task list owner: %w", err)
+		}
+
+		e.logger.Info(fmt.Sprintf(">> Debugging tasklist ownership : %v, %v", taskListOwner.Identity(), self.Identity()))
+		if taskListOwner.Identity() != self.Identity() {
+			toShutDown = append(toShutDown, manager)
+		}
+	}
+
+	e.logger.Info("Got list of non-owned-tasklists",
+		tag.Dynamic("tasklist-debug-info", toShutDown),
+	)
+	return toShutDown, nil
+}
+
+func (e *matchingEngineImpl) guardAgainstShardLoss(taskList *tasklist.Identifier) error {
```

Comment:
maybe `errIfShardLost` instead?  "guard" makes it sound to me like it's "doing" something, but it's just returning a value that has to be handled correctly elsewhere